internal/pkg/agent/application/enroll/enroll.go

```diff
@@ -0,0 +1,574 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package enroll
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"math/rand/v2"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/otiai10/copy"
+	"gopkg.in/yaml.v2"
+
+	"github.com/elastic/elastic-agent-libs/file"
+	"github.com/elastic/elastic-agent-libs/transport/tlscommon"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/filelock"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/info"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/paths"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/configuration"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
+	"github.com/elastic/elastic-agent/internal/pkg/config"
+	"github.com/elastic/elastic-agent/internal/pkg/core/backoff"
+	monitoringConfig "github.com/elastic/elastic-agent/internal/pkg/core/monitoring/config"
+	"github.com/elastic/elastic-agent/internal/pkg/crypto"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	fleetclient "github.com/elastic/elastic-agent/internal/pkg/fleetapi/client"
+	"github.com/elastic/elastic-agent/internal/pkg/release"
+	"github.com/elastic/elastic-agent/internal/pkg/remote"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+const (
+	EnrollBackoffInit = time.Second * 5
+	EnrollBackoffMax  = time.Minute * 10
+
+	maxRetriesstoreAgentInfo       = 5
+	defaultFleetServerHost         = "0.0.0.0"
+	defaultFleetServerPort         = 8220
+	defaultFleetServerInternalHost = "localhost"
+	defaultFleetServerInternalPort = 8221
+	statusPath                     = "/api/status"
+	apiStatusTimeout               = 15 * time.Second
+	backupSuffix                   = ".enroll.bak"
+)
+
+type saver interface {
+	Save(io.Reader) error
+}
+
+func CheckRemote(ctx context.Context, c fleetclient.Sender) error {
```

Comment:
Nit: Could we move this out of the `enroll` package, seeing that it doesn't do anything specific to enrollment and also that it's called from the policy change handler code?